libraries/integration/Microsoft.Bot.Builder.Integration.Runtime/Extensions/ServiceCollectionExtensions.cs

```diff
@@ -0,0 +1,235 @@
+        /// <param name="configuration">The application configuration.</param>
+        public static void AddBotRuntime(this IServiceCollection services, IConfiguration configuration)
+        {
+            if (services == null)
+            {
+                throw new ArgumentNullException(nameof(services));
+            }
+
+            if (configuration == null)
+            {
+                throw new ArgumentNullException(nameof(configuration));
+            }
+
+            // Component registrations must be added before the resource explorer is instantiated to ensure
+            // that all types are correctly registered. Any types that are registered after the resource explorer
+            // is instantiated will not be picked up otherwise.
+            ComponentRegistrations.Add();
+
+            // Configuration
+            string applicationRoot = configuration.GetSection(ConfigurationConstants.ApplicationRootKey).Value;
+            string defaultLocale = configuration.GetSection(ConfigurationConstants.DefaultLocale).Value;
+            string rootDialog = configuration.GetSection(ConfigurationConstants.RootDialogKey).Value;
+            
+            // Runtime settings. If no config is provided, we create basic runtime config with defaults.
+            var runtimeSettings = configuration.GetSection(ConfigurationConstants.RuntimeSettingsKey).Get<RuntimeSettings>() ?? new RuntimeSettings();
+
+            // Bot
+            services.AddSingleton<IBot, CoreBot>();
+            services.AddOptions()
+                .Configure<CoreBotOptions>(o =>
+                {
+                    o.DefaultLocale = defaultLocale;
+                    o.RootDialog = rootDialog;
+                });
+
+            services.AddSingleton(configuration);
+
+            // ResourceExplorer. TryAddSingleton will only add if there is no other registration for resource explorer.
+            // Tests use this to inject custom resource explorers but could also be used for advanced runtime customization scenarios.
+            services.TryAddSingleton<ResourceExplorer>(serviceProvider =>
+                new ResourceExplorer()
+                    .AddFolder(applicationRoot)
+                    .RegisterType<OnQnAMatch>(OnQnAMatch.Kind));
+            
+            // Runtime set up
+            services.AddBotRuntimeSkills(runtimeSettings.Skills);
+            services.AddBotRuntimeStorage(configuration, runtimeSettings);
+            services.AddBotRuntimeTelemetry(runtimeSettings.Telemetry);
+            services.AddBotRuntimeTranscriptLogging(configuration, runtimeSettings.Features);
+            services.AddBotRuntimeFeatures(runtimeSettings.Features);
+            services.AddBotRuntimePlugins(configuration, runtimeSettings);
+            services.AddBotRuntimeAdapters(runtimeSettings);
+        }
+
+        internal static void AddBotRuntimeSkills(this IServiceCollection services, SkillSettings skillSettings)
```

Comment:
I love these internal methods...

Reply:
Yes, the question is (for R13 when this goes GA): do we want a big black box that adds the runtime (and has well documented ways of overriding pieces like adding your own storage singleton or bot or adapter by modifying startupcs?). Or do we split this in to a few black boxes that make sense independently?

Otherwise, the current composer runtime is a full white box but folks wanted to move away from that: https://github.com/microsoft/BotFramework-Composer/blob/main/runtime/dotnet/azurewebapp/Startup.cs

The DI logic is more complicated than samples because samples do only one thing, while this does different things based on config. 

With all of that in mind, any proposals for what the programming experience would look like in an ideal world? I also think that the public 'AddBotRuntime' is probably too big of a black box. However I still have not landed with a fantastic improvement, but I would *love* to ship R13 with another round of thinking about this specific experience and how the template startup.cs looks like.

Reply:
The problem should be solved in the design of the components not worked around in their assembly.